### PR TITLE
Add bitmask types to descr

### PR DESCRIPTION
**Pull request by josevalim**
The bitmap represents all non-divisible types.
Currently it implements all types in our system,
but atoms, lists, tuples, maps, and functions
will be given more precise types later on.

Dynamic has not been yet implemented nor a
optimizations for term.

### CODE REVIEW

**Review by gldubc — APPROVED** [review]

File: lib/elixir/lib/module/types/descr.ex
Comment (gldubc) [inline lib/elixir/lib/module/types/descr.ex:1]: I'd also add negation implemented as difference with term.
```
@doc """
Computes the negation of a type.
"""
def negation(%{} = descr), do: difference(term(), descr)
```

**Review by gldubc — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: lib/elixir/test/elixir/module/types/descr_test.exs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  describe "difference" do
+    test "bitmap" do
+      assert difference(float(), integer()) == float()
+      assert difference(union(float(), integer()), integer()) == float()
+      assert difference(union(float(), integer()), binary()) == union(float(), integer())
+    end
+
+    test "term" do
+      assert difference(float(), term()) == none()
+      assert difference(float(), term()) == none()
```
Comment (wojtekmach) [inline lib/elixir/test/elixir/module/types/descr_test.exs:52]: double-checking, is this clause is duplicated on purpose?

Comment (josevalim) [reply lib/elixir/test/elixir/module/types/descr_test.exs:52]: Good find, thank you!

**Pull request merged**